### PR TITLE
Only send flushes when Downstairs is idle; send Barrier otherwise

### PR DESCRIPTION
**Pull request by mkeeter**
This PR removes automatic flushes, per [RFD 518](https://rfd.shared.oxide.computer/rfd/518).  Instead, the new `Barrier` operation is sent.  If the system is idle for a particular amount of time, we send a final `Flush` to put everything into a known state.

When the Upstairs retires jobs after a barrier operation, the system as a whole becomes ineligible for replay.  This state determines whether the new Downstairs reconnects through `Offline` (which does replay) or `Faulted` (which does live-repair instead).

Removing automatic flushes is a noticeable performance improvement:

Screenshot 2024-10-11 at 11 06 26 AM

(tested on the London mini-cluster, with Upstairs and 3x Downstairs on different sleds)

```
region info:
  block size:      4096 bytes
  blocks / extent: 16384
  extent size:     64 MiB
  extent count:    2048
  total blocks:    33554432
  total size:      128 GiB
  encryption:      yes
```

### CODE REVIEW

**Comment by mkeeter** [comment]
Rebased to stage on top of #1507, because we want to only send `Barrier` operations if there's enough bytes / jobs buffered (which requires keeping track!)

File: upstairs/src/dummy_downstairs_tests.rs [hunk 1]
```diff
@@ -1473,6 +1528,17 @@ async fn test_byte_fault_condition() {
...
+
+        // Once one of the Downstairs has cached more that a certain amount of
+        // bytes, it will automatically send a Barrier operation.
+        if i == barrier_point {
+            let m = harness.ds1().recv().await.unwrap();
+            if !matches!(m, Message::Barrier { .. }) {
+                panic!("expected Barrier, got message {m:?}");
+            }
+            harness.ds2.ack_barrier().await;
+            harness.ds3.ack_barrier().await;
```
Comment (leftwo) [inline upstairs/src/dummy_downstairs_tests.rs:1540]: Do we have any way to check that replay is no longer available to this downstairs?  I'm not sure if the test has access at this layer to it.

**Comment by leftwo** [comment]
With this, we can probably close https://github.com/oxidecomputer/crucible/issues/1358 as fixed

File: upstairs/src/dummy_downstairs_tests.rs (hunk 1 above)
Comment (mkeeter) [reply upstairs/src/dummy_downstairs_tests.rs:1540]: The end of this unit test is checking that live-repair works, so I think we're implicitly testing that we don't do replay.

Comment (leftwo) [reply upstairs/src/dummy_downstairs_tests.rs:1540]: Ah, I was thinking more that the sending of a barrier operation has also flipped the `can_replay` in the `Downstairs` struct.  I'm not sure if we can do that though here (or in a test elsewhere), as `can_replay` might not be exposed like that.

Comment (mkeeter) [reply upstairs/src/dummy_downstairs_tests.rs:1541]: Yeah, it's not easy to check from the outside (when we only have the `Guest` handle).  If you really want it, we could probably add it to the `downstairs_state` helper function.

File: upstairs/src/downstairs.rs [hunk 2]
```diff
@@ -2474,6 +2538,13 @@ impl Downstairs {
                 "downstairs failed, too many outstanding bytes {byte_count}"
             );
             Some(ClientStopReason::TooManyOutstandingBytes)
+        } else if !self.can_replay {
+            // XXX can this actually happen?
+            warn!(
+                self.log,
+                "downstairs became ineligible for replay while offline"
```
[the remark this answers is not in the document]
Comment (mkeeter) [reply upstairs/src/downstairs.rs:2535]: I don't think it's possible, but the state machines are complex enough that I'm not 100% certain; faulting the downstairs is a correct behavior, so I'd rather do that versus panicking.  I did try running unit tests with a panic here, and it didn't fire, but that could just mean that it's an unlikely corner case.

> It also seems like the statement downstairs became ineligible for replay while offline is not the unusual part, is it? If a downstairs is Offline and we continue to get IO, then it will become ineligible for replay and that's kind of the expected behavior, though maybe not in this method?

The `can_replay` flag is set to `false` when we remove _completed_ jobs that would be needed for a replay ("completed" = done / skipped / error on all three downstairs).  That shouldn't happen with one downstairs Offline, because jobs are left `InProgress` for the offline downstairs (for exactly this reason).

(We can also become ineligible for replay when too many `InProgress` jobs pile up, which is checked right here)

File: upstairs/src/dummy_downstairs_tests.rs (hunk 1 above)
Comment (leftwo) [reply upstairs/src/dummy_downstairs_tests.rs:1541]: I don't think it's worth adding it to the `downstairs_state` helper function. 
I don't see anywhere else that seems like a good place either.  Nothing in the upstairs tests cover this case.

**Review by leftwo — APPROVED** [review]

**Review by faithanalog — APPROVED** [review]
We should probably tighten up `IO_CACHED_MAX_BYTES` soon. Right now we can theoretically buffer up to 1GiB of jobs for replay, and that feels like quite a lot to me. We usually won't- we will only get to that point if the guest is doing large amounts of IO, continuously so we never go idle, and the guest never sends a flush. That's unlikely/rare during normal filesystem operations (but will be easier to hit when writing to the raw block device like iodriver does). Because of that, we shouldn't expect more than a few VMs to hit this under normal operation. But in interest in not overcommitting resources, I think that bound should be lower.

**Review by jmpesp — APPROVED** [review]
